functions-continued/variables-are-references.md

```diff
@@ -0,0 +1,648 @@
+
+## Extending the Concept: Nested Data Structures
+
+Lists and dictionaries are not only mutable data types, but they're also _containers_. As containers, they contain either mutable or immutable data types itself!
+
+The same principles extend to nested data structures.
+
+Consider the following questions:
+
+- Imagine a list within a list. If we modify the inner list, does that affect the references pointing to it?
+- Imagine a dictionary within a list. What are the effects of modifying the inner dictionary?
+
+Make a prediction. Run an experiment.
+
+<!-- Question  -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: paragraph
+* id: tRHUpg
+* title: Variables Are References
+##### !question
+
+Predict:
+
+- Imagine a list within a list. If we modify the inner list, does that affect the references pointing to it?
+- Imagine a dictionary within a list. What are the effects of modifying the inner dictionary?
+
+##### !end-question
+##### !placeholder
+
+I predict...
+
+##### !end-placeholder
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+Now, read through this code and its example output:
+
+```python
+peaches = ["Hello", [False, False, False] ,"World!"]
+
+pears = peaches[1]
+
+pears.append(True)
+
+print("value of peaches:", peaches)
+```
+
+```
+value of peaches: ['Hello', [False, False, False, True], 'World!']
+```
+
+`peaches[1]` _and_ `pears` both point to the mutable object `[False, False, False]`.
+
+When we modify that object with `pears.append()`, we modify the same object referred to by `peaches[1]`.
+
+Let's look at an example dictionary inside a list:
+
+```python
+pears = {
+    "hello": "world!"
+}
+
+peaches = [False, pears, False]
+
+pears["hello"] = "modified value"
+
+print("value of peaches:", peaches)
+```
+
+```
+value of peaches: [False, {'hello': 'modified value'}, False]
+```
+
+`pears` and `peaches[1]` refer to the same dictionary object.
+
+Giving a new value to a key in a dictionary counts as modification. Modifying `pear` affects the item at `peaches[1]`.
+
+### !callout-info
+
+## Endless Combinations
+
+Curious about more combinations, such as a list in a dictionary, or a list inside a list inside a list? Follow your curiosity and experiment!
+
+### !end-callout
+
+### !callout-success
+
+## Foreshadowing: OOP
+
+When discussing object-oriented programming, we'll write user-defined classes. User-defined classes are mutable data types by default.
+
+### !end-callout
+
+## Best Pattern: Reassign Often, Return Often
+
+References and immutability should primarily affect our programming with awareness.
+
+Without awareness, modifying a mutable object risks unintentionally affecting other references.
+
+As general patterns to takeaway:
+
+1. Stay aware of the difference between modifying and reassigning
+1. While programming, create diagrams to help understand references between variables and objects
+1. It's more reliable to _return_ any important values from a function, rather than modify a value
```

Comment:
Provide justification for why we should return results rather than modifying inputs.

```suggestion
1. It's usually more desirable to _return_ important results from a function, rather than modifying a value that was passed in
    1. Callers of our functions usually don't expect the values they pass in to be modified unless there is clear documentation that your function will do so
```